File: lib/extorch/jit/jit.ex
defmodule ExTorch.JIT do
  @moduledoc """
  TorchScript model loading, inference, and management.

  This module provides functions to load pre-trained TorchScript (`.pt`) models,
  run inference, inspect model structure, and manage model lifecycle.

  ## Example

      model = ExTorch.JIT.load("model.pt")
      input = ExTorch.randn({1, 3, 224, 224})
      output = ExTorch.JIT.forward(model, [input])

  """

  alias ExTorch.JIT.Model

  @doc """
  Load a TorchScript model from a file.

  ## Arguments
    - `path`: Path to the `.pt` file.
    - `opts`: Keyword list of options.
      - `:device` - Device to load the model onto (default: `:cpu`).

  ## Returns
  A `%ExTorch.JIT.Model{}` struct.

  ## Examples

      model = ExTorch.JIT.load("model.pt")
      model = ExTorch.JIT.load("model.pt", device: {:cuda, 0})
  """
  @spec load(String.t(), keyword()) :: Model.t()
  def load(path, opts \\ []) do
    device = Keyword.get(opts, :device, :cpu)
    ExTorch.Native.jit_load(path, device)
  end

  @doc """
  Save a TorchScript model to a file.

  ## Arguments
    - `model`: The model to save.
    - `path`: Path to save the `.pt` file.
  """
  @spec save(Model.t(), String.t()) :: :ok
  def save(%Model{} = model, path) do
    ExTorch.Native.jit_save(model, path)
    :ok
  end

  @doc """
  Run the forward method on a model.

  ## Arguments
    - `model`: The loaded model.
    - `inputs`: A list of input tensors.

  ## Returns
  The model output, which can be a tensor, tuple, list, map, or scalar
  depending on the model's return type.

  ## Examples

      input = ExTorch.randn({1, 784})
      output = ExTorch.JIT.forward(model, [input])
  """
  @spec forward(Model.t(), [ExTorch.Tensor.t()]) :: term()
  def forward(%Model{} = model, inputs) when is_list(inputs) do
    ExTorch.Native.jit_forward(model, inputs)
  end

  @doc """
  Invoke a named method on a model.

  ## Arguments
    - `model`: The loaded model.
    - `method_name`: Name of the method to invoke.
    - `inputs`: A list of input tensors.

  ## Returns
  The method output.
  """
  @spec invoke(Model.t(), String.t(), [ExTorch.Tensor.t()]) :: term()
  def invoke(%Model{} = model, method_name, inputs \\ []) when is_list(inputs) do
    ExTorch.Native.jit_invoke_method(model, method_name, inputs)
  end

  @doc """
  Get the names of all methods on a model.

  ## Returns
  A list of method name strings.
  """
  @spec methods(Model.t()) :: [String.t()]
  def methods(%Model{} = model) do
    ExTorch.Native.jit_get_method_names(model)
  end

  @doc """
  Get named parameters of a model.

  ## Returns
  A list of `{name, tensor}` tuples.
  """
  @spec parameters(Model.t()) :: [{String.t(), ExTorch.Tensor.t()}]
  def parameters(%Model{} = model) do
    ExTorch.Native.jit_named_parameters(model)
  end

  @doc """
  Get named buffers of a model.

  ## Returns
  A list of `{name, tensor}` tuples.
  """
  @spec buffers(Model.t()) :: [{String.t(), ExTorch.Tensor.t()}]
  def buffers(%Model{} = model) do
    ExTorch.Native.jit_named_buffers(model)
  end

  @doc """
  Get names of submodules of a model.

  ## Returns
  A list of submodule name strings.
  """
  @spec modules(Model.t()) :: [String.t()]
  def modules(%Model{} = model) do
    ExTorch.Native.jit_named_modules(model)
  end

  @doc """
  Set a model to evaluation mode.
  """
  @spec eval(Model.t()) :: :ok
  def eval(%Model{} = model) do
    ExTorch.Native.jit_set_eval(model)
    :ok
  end

  @doc """
  Set a model to training mode.
  """
  @spec train(Model.t()) :: :ok
  def train(%Model{} = model) do
    ExTorch.Native.jit_set_train(model)
    :ok
  end

  @doc """
  Move a model to a different device.

  ## Arguments
    - `model`: The model to move.
    - `device`: Target device (e.g., `:cpu`, `{:cuda, 0}`).

  ## Returns
  A new `%ExTorch.JIT.Model{}` on the target device.
  """
  @spec to(Model.t(), ExTorch.Device.device()) :: Model.t()
  def to(%Model{} = model, device) do
    ExTorch.Native.jit_to_device(model, device)
  end
end

File: native/extorch/src/nifs/jit.rs
use crate::encoding::jit::{ivalue_flat_to_term, named_tensors_to_term};
use crate::native::torch;
use crate::shared_types::{JitModuleStruct, Reference, TensorStruct};

use rustler::{Atom, Encoder, Env, Error, NifResult, ResourceArc, Term};

/// Build a TensorList from a slice of TensorStructs.
fn make_tensor_list(inputs: &[TensorStruct]) -> torch::TensorList {
    let values: Vec<torch::TensorOut> = inputs
        .iter()
        .map(|t| torch::TensorOut {
            tensor: t.resource.tensor.clone(),
            used: true,
        })
        .collect();
    torch::TensorList { values, used: true }
}

/// Helper to convert a cxx error into a NifResult error.
fn cxx_err_to_nif(err: cxx::Exception) -> Error {
    let err_msg = err.what().to_owned();
    let err_parts: Vec<&str> = err_msg.split('\n').collect();
    Error::RaiseTerm(Box::new(err_parts[0].to_owned()))
}

/// Clone a cxx Device struct (cxx shared structs don't auto-derive Clone).
fn clone_device(device: &torch::Device) -> torch::Device {
    torch::Device {
        device: device.device.clone(),
        index: device.index,
    }
}

/// Load a TorchScript model from a file path.
#[rustler::nif]
pub fn jit_load<'a>(
    path: String,
    device: torch::Device,
) -> NifResult<JitModuleStruct<'a>> {
    let elixir_device = clone_device(&device);
    let module = torch::jit_load(path, device).map_err(cxx_err_to_nif)?;
    let wrapped = torch::CrossModuleRef { module };
    let resource = ResourceArc::new(wrapped);
    Ok(JitModuleStruct {
        resource,
        reference: Reference::new(),
        device: elixir_device,
    })
}

/// Save a TorchScript model to a file path.
#[rustler::nif]
pub fn jit_save(model: JitModuleStruct, path: String) -> NifResult<()> {
    torch::jit_save(&model.resource.module, path).map_err(cxx_err_to_nif)
}

/// Run the forward method on a JIT model, returning an Elixir term.
#[rustler::nif]
pub fn jit_forward<'a>(
    env: Env<'a>,
    model: JitModuleStruct<'a>,
    inputs: Vec<TensorStruct<'a>>,
) -> NifResult<Term<'a>> {
    let input_list = make_tensor_list(&inputs);
    let flat = torch::jit_forward(&model.resource.module, input_list).map_err(cxx_err_to_nif)?;
    Ok(ivalue_flat_to_term(env, &flat))
}

/// Invoke a named method on a JIT model.
#[rustler::nif]
pub fn jit_invoke_method<'a>(
    env: Env<'a>,
    model: JitModuleStruct<'a>,
    method_name: String,
    inputs: Vec<TensorStruct<'a>>,
) -> NifResult<Term<'a>> {
    let input_list = make_tensor_list(&inputs);
    let flat = torch::jit_invoke_method(&model.resource.module, method_name, input_list)
        .map_err(cxx_err_to_nif)?;
    Ok(ivalue_flat_to_term(env, &flat))
}

/// Get names of all methods on a JIT model.
#[rustler::nif]
pub fn jit_get_method_names(model: JitModuleStruct) -> NifResult<Vec<String>> {
    let names = torch::jit_get_method_names(&model.resource.module).map_err(cxx_err_to_nif)?;
    Ok(names.into_iter().map(|s| s.to_string()).collect())
}

/// Get named parameters of a JIT model as a list of {name, tensor} tuples.
#[rustler::nif]
pub fn jit_named_parameters<'a>(env: Env<'a>, model: JitModuleStruct<'a>) -> NifResult<Term<'a>> {
    let params =
        torch::jit_named_parameters(&model.resource.module).map_err(cxx_err_to_nif)?;
    Ok(named_tensors_to_term(env, &params))
}

/// Get named buffers of a JIT model as a list of {name, tensor} tuples.
#[rustler::nif]
pub fn jit_named_buffers<'a>(env: Env<'a>, model: JitModuleStruct<'a>) -> NifResult<Term<'a>> {
    let buffers =
        torch::jit_named_buffers(&model.resource.module).map_err(cxx_err_to_nif)?;
    Ok(named_tensors_to_term(env, &buffers))
}

/// Get names of submodules of a JIT model.
#[rustler::nif]
pub fn jit_named_modules(model: JitModuleStruct) -> NifResult<Vec<String>> {
    let names = torch::jit_named_modules(&model.resource.module).map_err(cxx_err_to_nif)?;
    Ok(names.into_iter().map(|s| s.to_string()).collect())
}

/// Set a JIT model to evaluation mode.
#[rustler::nif]
pub fn jit_set_eval(model: JitModuleStruct) -> NifResult<()> {
    torch::jit_set_eval(&model.resource.module).map_err(cxx_err_to_nif)
}

/// Set a JIT model to training mode.
#[rustler::nif]
pub fn jit_set_train(model: JitModuleStruct) -> NifResult<()> {
    torch::jit_set_train(&model.resource.module).map_err(cxx_err_to_nif)
}

/// Move a JIT model to a different device.
#[rustler::nif]
pub fn jit_to_device<'a>(
    model: JitModuleStruct<'a>,
    device: torch::Device,
) -> NifResult<JitModuleStruct<'a>> {
    let elixir_device = clone_device(&device);
    let module =
        torch::jit_to_device(&model.resource.module, device).map_err(cxx_err_to_nif)?;
    let wrapped = torch::CrossModuleRef { module };
    let resource = ResourceArc::new(wrapped);
    Ok(JitModuleStruct {
        resource,
        reference: Reference::new(),
        device: elixir_device,
    })
}

// ============================================================================
// IR Introspection
// ============================================================================

/// Get the JIT graph IR as a string.
#[rustler::nif]
pub fn jit_graph_str(model: JitModuleStruct) -> NifResult<String> {
    let graph = torch::jit_graph_str(&model.resource.module).map_err(cxx_err_to_nif)?;
    Ok(graph.to_string())
}

/// Get parameter info for a JIT module.
/// Returns a list of maps: %{name, shape, dtype, requires_grad}
#[rustler::nif]
pub fn jit_module_parameters_info<'a>(env: Env<'a>, model: JitModuleStruct<'a>) -> NifResult<Term<'a>> {
    let params = torch::jit_module_parameters_info(&model.resource.module).map_err(cxx_err_to_nif)?;

    let result: Vec<Term<'a>> = params.iter().map(|p| {
        let name: &str = p.name.as_str();
        let shape: Vec<i64> = p.shape.iter().copied().collect();
        let dtype_str: &str = p.dtype.as_str();
        let dtype_atom = Atom::from_str(env, dtype_str).unwrap();

        let keys = vec![
            Atom::from_str(env, "name").unwrap().encode(env),
            Atom::from_str(env, "shape").unwrap().encode(env),
            Atom::from_str(env, "dtype").unwrap().encode(env),
            Atom::from_str(env, "requires_grad").unwrap().encode(env),
        ];
        let values = vec![
            name.encode(env),
            shape.encode(env),
            dtype_atom.encode(env),
            p.requires_grad.encode(env),
        ];

        Term::map_from_arrays(env, &keys, &values).unwrap()
    }).collect();

    Ok(result.encode(env))
}

/// Get submodule info for a JIT module.
/// Returns a list of maps: %{name, type_name, parameters}
#[rustler::nif]
pub fn jit_module_submodules_info<'a>(env: Env<'a>, model: JitModuleStruct<'a>) -> NifResult<Term<'a>> {
    let submodules = torch::jit_module_submodules_info(&model.resource.module).map_err(cxx_err_to_nif)?;

    let result: Vec<Term<'a>> = submodules.iter().map(|s| {
        let name: &str = s.name.as_str();
        let type_name: &str = s.type_name.as_str();

        let params: Vec<Term<'a>> = s.parameters.iter().map(|p| {
            let pname: &str = p.name.as_str();
            let shape: Vec<i64> = p.shape.iter().copied().collect();
            let dtype_str: &str = p.dtype.as_str();
            let dtype_atom = Atom::from_str(env, dtype_str).unwrap();

            let keys = vec![
                Atom::from_str(env, "name").unwrap().encode(env),
                Atom::from_str(env, "shape").unwrap().encode(env),
                Atom::from_str(env, "dtype").unwrap().encode(env),
                Atom::from_str(env, "requires_grad").unwrap().encode(env),
            ];
            let values = vec![
                pname.encode(env),
                shape.encode(env),
                dtype_atom.encode(env),
                p.requires_grad.encode(env),
            ];

            Term::map_from_arrays(env, &keys, &values).unwrap()
        }).collect();

        let keys = vec![
            Atom::from_str(env, "name").unwrap().encode(env),
            Atom::from_str(env, "type_name").unwrap().encode(env),
            Atom::from_str(env, "parameters").unwrap().encode(env),
        ];
        let values = vec![
            name.encode(env),
            type_name.encode(env),
            params.encode(env),
        ];

        Term::map_from_arrays(env, &keys, &values).unwrap()
    }).collect();

    Ok(result.encode(env))
}

/// Get method names for a JIT module.
#[rustler::nif]
pub fn jit_module_methods_info(model: JitModuleStruct) -> NifResult<Vec<String>> {
    let methods = torch::jit_module_methods_info(&model.resource.module).map_err(cxx_err_to_nif)?;
    Ok(methods.into_iter().map(|s| s.to_string()).collect())
}
